### DISCUSSION TITLE
Recommended production docker setup

### DISCUSSION THREAD

**Discussion by jcpprojects** [docker-library/wordpress#934] [General]
What is the recommended docker setup for production if you want to have several Docker instances?

From here:
https://developer.wordpress.org/advanced-administration/security/backup/files
and here:
https://github.com/docker-library/wordpress/issues/924

Seems we need to persist /var/www/html

But then do we need a shared volume with all the code to be able to run more than one machine? Isn't this going to impact performance as some other posts suggest due to slower file access?

Some other social media posts suggest to only persist wp-content/uploads directory, and core installation and plugins would come from Docker. I think should work, but then I am not sure about other local files which might be lost during updates which could be used by plugins, or .htaccess mentioned on the first link.

On the link below this solution is also suggested and used with mixed opinions (external db, persist wp-content/updates and replace the rest on every update), but I cannot see an official documentation supporting this.

https://www.reddit.com/r/docker/comments/h95lxv/does_anyone_run_wordpress_in_production_via/

**Comment by tianon** [comment]
How to set this up for your use case depends a lot on what you plan to do with it.  If you plan to use WordPress in the "classical" way where it auto-updates itself, plugins are installed via the web UI, etc, then your options are pretty limited (you'll need to do something to ensure only one instance ever attempts the auto-upgrades and you'll need to have some kind of shared storage across your instances for all of the WordPress installation directory, as you've described).

If, however, you have a mostly static instance or only care about a fixed set of plugins/themes and want to perform upgrades via Docker image rebuilds, then you should be able to get away with something more targeted like only sharing `wp-content/uploads` between instances.  You'll also want to make sure the rest of the installation is fully read-only so that you'll know quickly if something you're trying to do (plugins, etc) is writing to a location you haven't persisted/synchronized.

If you set your container's `WORKDIR` to `/usr/src/wordpress`, it _should_ avoid the automatic copying behavior and read the original image copy of the files too, which will improve your startup times and your ability to make sure that entire directory is read-only (`--read-only` on your containers).

I think the most important thing to keep in mind is that the automatic upgrades exist primarily for security updates, so you'll want to make sure whatever you set up has a quick redeployment time and that you're proactive about applying updates.

**Reply by jcpprojects** [reply]
In the case of a single machine or shared drive, I understand the auto-upgrades are managed with the .maintenance file which WP writes during upgrades, which works as exclusive lock if you use a shared drive or single machine. 

- Is this "classic" setup the recommended approach -shared drive and do not ever update Docker again though docker but let WP manage it-? 
- Does WP perform well with a shared drive? 
- What kind of file shares solutions are recommended? Thinking about the different clould providers.

Is docker used much in production for WP? I find it very useful for development, but haven't found many resources when it comes to realistic production like setups.

**Reply by tianon** [reply]
I use it in production, but not with more than a single instance.

You'd still want to keep your image up-to-date so that you get PHP (security) updates, but yeah, you wouldn't get WordPress via image updates anymore.

I don't know about which shared drive solutions are "recommended" -- I don't think Docker really changes anything here, and that's more of a general WP question (since the same would apply in the same way without Docker involved; we're really replicating a "standard" install as much as possible).
